### ISSUE TITLE
PHPMD complaining about UnusedFormalParameters when overriding method from base class.

### ISSUE THREAD

**Issue by krtek4** [phpmd/phpmd#180]
Following issue #50, I also have a warning about UnusedFormalParameters which I think should no be happening.

I created classes extending the base Symfony 2.0 Command class which includes an `execute` method :

```
class Command
{
    [...]
    protected function execute(InputInterface $input, OutputInterface $output)
    {
        throw new \LogicException('You must override the execute() method in the concrete command class.');
    }
    [...]
}
```

My own classes are as follow : 

```
class BaseCommand extends Command
{
    [...]
}

class ConfigCommand extends BaseCommand
{
    protected function configure()
    {
        [...]
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $this->loadContext();
        $output->writeln('<info>Current configuration is:</info>');
        $output->writeln(Yaml::dump(Context::getInstance()->getParam('config')));
    }
}
```

PHPMD is complaining about the `$input` parameter that is not used on the `ConfigCommand`, but there is nothing I can do about it since I really don't need it and I also cannot obviously call the parent method or redefine it since it is part of the Symfony2 framework.

Thanks for your great tool !

**Comment by ravage84** [comment]
Duplicate of https://github.com/phpmd/phpmd/issues/24
But going to close the other one instead.

**Comment by manuelpichler** [comment]
We must investigate how this can be implemented in phpmd efficiently

**Comment by charles-martin** [comment]
FYI: This also applies to closures as well. IE, in the Phalcon Framework, it is common to do:
```php
DI::getDefault()->get('eventsManager')->attach(
    'some:event',
    function($event, $data) {
        $data->something = 'something-else';
    }
);
```
In this case, phpmd complains about the unused variable $event. It's not used, but it has to be there due to the format for the callbacks.

That said, I would say there should probably be two 'options' on this: 1) Ignore In Child Methods, and 2) Ignore Initial Variables If Second Variable Used.

**Comment by mikesherov** [comment]
I'm taking a stab at this right now. Thanks again for this excellent project.

**Comment by mikesherov** [comment]
Pull request here: https://github.com/phpmd/phpmd/pull/331 please let me know if we should be going in a different direction.

**Comment by derrabus** [comment]
Allowing unused parameters before used doesn't quite cut it. I can override a method without using any parameters of the base method, but I'm required to list them all in my method signature.

**Comment by mikesherov** [comment]
@derrabus, you have the `inheritdoc` workaround for that, no?

**Comment by derrabus** [comment]
@mikesherov I don't know what the "`inheritdoc` workaround" is. But a workaround is usually not a solution. :wink:

**Comment by mikesherov** [comment]
@derrabus, as PHPMD only analyzes individual files, and as PHP has no canonical way to specify a function as overriding an inherited methods, PHPMD relies on the presence of the `inheritdoc` notation of `phpdoc` in the methods docblock.

**Comment by derrabus** [comment]
I see. I'll try that. But for analyzing generic code that I did not write, that workaround won't solve the problem.

**Comment by mikesherov** [comment]
You're running PHPMD on 3rd party code? Yeah, nothing is going to help you there. The typically use case for PHPMD is analyzing the source code in your control.

**Comment by derrabus** [comment]
Well, I said I did not write that code. I didn't say the code is not in my control. It just wasn't at the time of writing. I already know that my code is good, so I'm running this tool on other people's code, you know? :wink:

But that discussion does not belong here. I'm running phpmd to analyze code, which imho is the main purpose of this tool. And this issue generates quite a few false positives. I understand that, because of the way PHPMD works right now, this issue is hard to fix and I won't complain about this. All I'm saying is that your pull request is not solving this issue, just a special case of it.

**Comment by mikesherov** [comment]
It's unlikely PHPMD will solve this in any way other than relying on comments (which it does) and special casing positional parameters because to be able to truly detect that a method is being overridden, PHPMD would need to execute your script so that it could determine your autoloading strategy and the source file of the class you're inheriting from (and also all the source files all the way up the class hierarchy). As PHPMD is a static analysis tool (rules about the syntax and structure of your code that can be determined without execution), this will likely never happen.

**Comment by derrabus** [comment]
> PHPMD would need to execute your script so that it could determine your autoloading strategy and the source file of the class you're inheriting from (and also all the source files all the way up the class hierarchy).

I'd disagree on this one. For instance, IDEs like PhpStorm able to determine inheritance without knowing my autoloading strategy or executing my code. It actually is possible to do this in static analysis, but not if the code is analyzed on a by-file basis.

Again: I know that this is not easy to fix right now and I accept that. All I'm saying is that your PR does not fix this issue.

**Comment by ravage84** [comment]
@mikesherov many thanks for your attempt to solve this issue by introducing a new config property. Unfortunately I, too, don't think it's a good solution to the problem at hand.

I guess currently there are only one half hearted solution:
Suppressing the issue case by case by using the ``@SuppressWarnings(PHPMD.UnusedFormalParameter)`` annotation.

For the future, the ["Dynamic warning suppressions"](https://github.com/phpmd/phpmd/issues/30) by using parameterized annotations could minimize the issue even further:
That __could__ look like this then: ``@SuppressWarnings(PHPMD.UnusedFormalParameter("event"))``, where the parameter ``$event`` would be ignored and the other one, ``$data`` would be not.

Another - still quite half-hearted - solution could be to add an ``exceptions`` config parameter to all the ``Unused*`` rules, see https://github.com/phpmd/phpmd/pull/329 for example. This would be simple to implement, but I don't think it would be a good idea either as it would create many unwanted blind spots.

As for the inerhitance hierarchy, I don't believe we can get this done in an efficient and satisfiing manner. But neither me or @manuelpichler have looked into this, yet, due to lack of time.

@derrabus you can't compare PhpStorm and phpmd in a fair way, since PhpStorm does much more than phpmd does or could do. For example it does a full index of all files in your project folder (and in some cases beyond that), also it does some assumption about your autoloading or even gets configured to use some autoloading logic.
Anyway, I think this is beyond of what phpmd should do - at least for the foreseeable future.

To sum up: I think we should invest time and effort in implementing the ["Dynamic warning suppressions"](https://github.com/phpmd/phpmd/issues/30) as this would open the door to more fine-grained suppressions and thus would solve many of the issues people (could) have.

**Comment by mikesherov** [comment]
Thanks for the reply @ravage84

**Comment by ravage84** [comment]
> That could look like this then: ``@SuppressWarnings(PHPMD.UnusedFormalParameter("event"))``, where the parameter ``$event`` would be ignored and the other one, ``$data`` would be not.

Looking at this again, I still think this would be the best solution.

Alternatively, it could go in the direction described in https://github.com/phpmd/phpmd/issues/149#issuecomment-305939769.

Moving this issue to Future, though.

**Comment by ravage84** [comment]
Moving discussion to #515
